File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        slow = dummy
        fast = dummy

        # Move fast pointer n steps ahead
        for _ in range(n):
            fast = fast.next

        # Move both pointers until fasr reaches the end
        while fast.next:
            slow = slow.next
            fast = fast.next 
        
        # Remove the n-th node from the end
        slow.next = slow.next.next

        return dummy.next

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        // Pointer Manipulation: Each node only knows about next in a singly linked list, in a reverse node A -> B must become B -> A
        // State Tracking: Changing curr.next breaks the forward traversal, a temporary pointer should store curr.next before overwriting
        // 3 Pointer Tech: Keep prev(init. null), curr(init. head), and nextTemp to iteratively reverse pointers in a single pass w/o extra mem.
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null){
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }

        return prev;
        
    }
}
